src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+                Debug.Assert(firstInRegion != null);
+                Debug.Assert(lastInRegion != null);
+                int startLocation = firstInRegion.Syntax.SpanStart;
+                int endLocation = lastInRegion.Syntax.Span.End;
+                int length = endLocation - startLocation;
+                Debug.Assert(length >= 0, "last comes before first");
+                this.RegionSpan = new TextSpan(startLocation, length);
+            }
+
+            PendingBranches = ArrayBuilder<PendingBranch>.GetInstance();
+            _labelsSeen = PooledHashSet<BoundStatement>.GetInstance();
+            _labels = PooledDictionary<LabelSymbol, TLocalState>.GetInstance();
+            this.Diagnostics = DiagnosticBag.GetInstance();
+            this.compilation = compilation;
+            _member = member;
+            this.methodMainNode = node;
+            this.firstInRegion = firstInRegion;
+            this.lastInRegion = lastInRegion;
+            _loopHeadState = new Dictionary<BoundLoopStatement, TLocalState>(ReferenceEqualityComparer.Instance);
+            _trackRegions = trackRegions;
+            _trackExceptions = trackExceptions;
+            _nonMonotonicTransfer = nonMonotonicTransferFunction;
+        }
+
+        protected abstract string Dump(TLocalState state);
+
+        protected string Dump()
         {
-            this._trackUnassignments = trackUnassignments;
+            return IsConditionalState
+                ? $"true: {Dump(this.StateWhenTrue)} false: {Dump(this.StateWhenFalse)}"
+                : Dump(this.State);
         }
 
-        protected abstract void UnionWith(ref TLocalState self, ref TLocalState other);
+#if DEBUG
+        protected string DumpLabels()
+        {
+            StringBuilder result = new StringBuilder();
+            result.Append("Labels{");
+            bool first = true;
+            foreach (var key in _labels.Keys)
+            {
+                if (!first) result.Append(", ");
+                string name = key.Name;
+                if (string.IsNullOrEmpty(name)) name = "<Label>" + key.GetHashCode();
+                result.Append(name).Append(": ").Append(this.Dump(_labels[key]));
+                first = false;
+            }
+            result.Append("}");
+            return result.ToString();
+        }
+#endif
 
         /// <summary>
-        /// Nontrivial implementation is required for DataFlowsOutWalker or any flow analysis pass that "tracks
-        /// unassignments" like the nullable walker. The result should be a state, for each variable, that is
-        /// the strongest result possible (i.e. definitely assigned for the data flow passes, or not null for
-        /// the nullable analysis).  Slightly more formally, this should be a reachable state that won't affect
-        /// another reachable state when this is intersected with the other state.
+        /// Subclasses may override EnterRegion to perform any actions at the entry to the region.
```

Comment:
could you specify if subclasses must always call base.EnterRegion.  If not, then then having this be virtual is fine.  If so, you might want to consider making this non-virtual, but having it call out to a virtual no-op method htat subclasses override.  THe benefit there is then you don't have to worry about subclasses forgetting to call into the base.

Reply:
Virtually all of this code is simply copied from what used to be the superclass.  Changes to further improve the code are a good idea, but can just as well be done separately.

---
In reply to: [236565045](https://github.com/dotnet/roslyn/pull/31389#discussion_r236565045) [](ancestors = 236565045)

Reply:
Yes.  I unfortunately realized this late into the review.  I thought this was all new code to it all bein green.  I recommended to andy that this PR be changed to use partial classes in mupltiple files in order to help the it diff out.  With that, the diff wouldn't show this as all new code, but would instead show it as a file-rename with minimal changes.  Then, it would be easier to focus on what was changed.

Sorry for all the feedback noise, it was due to a misunderstanding of to what extent this change was, along with Andy's request for info on areas where documentation could be improved. I went through all the new code looking for things that weren't necessarily clear to me, even with a reasonable passing understanding of how flow-analysis generally works here.  THanks!